src/server/campaignd/server.cpp

```diff
@@ -944,6 +944,183 @@ void server::register_handlers()
+	REGISTER_CAMPAIGND_HANDLER(addon_downloads_by_version);
+	REGISTER_CAMPAIGND_HANDLER(forum_auth_usage);
+	REGISTER_CAMPAIGND_HANDLER(admins_list);
+	REGISTER_CAMPAIGND_HANDLER(hide_addon);
+	REGISTER_CAMPAIGND_HANDLER(unhide_addon);
+	REGISTER_CAMPAIGND_HANDLER(list_hidden);
+}
+
+void server::handle_list_hidden(const server::request& req)
+{
+	config response;
+	std::string username = req.cfg["username"].str();
+	std::string passphrase = req.cfg["passphrase"].str();
+
+	if(!authenticate_admin(username, passphrase)) {
+		send_error("The passphrase is incorrect.", req.sock);
+		return;
+	}
+
+	for(const auto& addon : addons_) {
+		if(addon.second["hidden"].to_bool()) {
+			config& child = response.add_child("hidden");
+			child["addon"] = addon.second["name"].str();
+		}
+	}
+
+	std::ostringstream ostr;
+	write(ostr, response);
+
+	const auto& wml = ostr.str();
+	simple_wml::document doc(wml.c_str(), simple_wml::INIT_STATIC);
+	doc.compress();
+
+	utils::visit([this, &doc](auto&& sock) { async_send_doc_queued(sock, doc); }, req.sock);
+}
+
+void server::handle_hide_addon(const server::request& req)
+{
+	std::string addon_id = req.cfg["addon"].str();
+	std::string username = req.cfg["username"].str();
+	std::string passphrase = req.cfg["passphrase"].str();
+
+	auto addon = get_addon(addon_id);
+
+	if(!addon) {
+		ERR_CS << "Add-on '" << addon_id << "' not found, cannot hide";
+		send_error("The add-on was not found.", req.sock);
+		return;
+	} else {
+		if(!authenticate_admin(username, passphrase)) {
+			send_error("The passphrase is incorrect.", req.sock);
+			return;
+		}
+
+		addon["hidden"] = true;
+		mark_dirty(addon_id);
+		write_config();
+		LOG_CS << "Add-on '" << addon_id << "' is now hidden";
+	}
+
+	send_message("Add-on hidden.", req.sock);
+}
+
+void server::handle_unhide_addon(const server::request& req)
+{
+	std::string addon_id = req.cfg["addon"].str();
+	std::string username = req.cfg["username"].str();
+	std::string passphrase = req.cfg["passphrase"].str();
+
+	auto addon = get_addon(addon_id);
+
+	if(!addon) {
+		ERR_CS << "Add-on '" << addon_id << "' not found, cannot unhide";
+		send_error("The add-on was not found.", req.sock);
+		return;
+	} else {
+		if(!authenticate_admin(username, passphrase)) {
+			send_error("The passphrase is incorrect.", req.sock);
+			return;
+		}
+
+		addon["hidden"] = false;
+		mark_dirty(addon_id);
+		write_config();
+		LOG_CS << "Add-on '" << addon_id << "' is now unhidden";
+	}
+
+	send_message("Add-on unhidden.", req.sock);
+}
+
+void server::handle_addon_count(const server::request& req)
+{
+	config response;
+	response["count"] = addons_.size();
+
+	std::ostringstream ostr;
+	write(ostr, response);
+
+	const auto& wml = ostr.str();
+	simple_wml::document doc(wml.c_str(), simple_wml::INIT_STATIC);
+	doc.compress();
+
+	utils::visit([this, &doc](auto&& sock) { async_send_doc_queued(sock, doc); }, req.sock);
+}
+
+void server::handle_addon_count_by_type(const server::request& req)
```

Comment:
This can be determined from the normal addon list that the client gets either way, no?

Reply:
Updated.